Bad cascading "is not a trait" error

## Meta

```
$ rustc -V
rustc 1.12.0-dev (5556554e2 2016-07-31)
```

## STR

```Rust
use std::Xxx;
struct Foo;
impl Xxx for Foo {}
fn main() {}
```

## Expected result

Single unresolved import error

## Actual result

```
error[E0432]: unresolved import `std::Xxx`. There is no `Xxx` in `std`
 --> <anon>:1:5
  |
1 | use std::Xxx;
  |     ^^^^^^^^

error[E0404]: `Xxx` is not a trait
 --> <anon>:3:6
  |
3 | impl Xxx for Foo {}
  |      ^^^

error: cannot continue compilation due to previous error
```

The E0404 error is a duplicate (BTW, E0404 and E0432 should be switched ;-)).

cc @jseyfried

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -1822,22 +1822,7 @@ impl<'a> Resolver<'a> {
 
             // If it's a typedef, give a note
             if let Def::TyAlias(..) = path_res.base_def {
-                let trait_name = trait_path.segments.last().unwrap().identifier.name;
-                err.span_label(trait_path.span, &format!("`{}` is not a trait", trait_name));
-
-                let definition_site = {
-                    let segments = &trait_path.segments;
-                    if trait_path.global {
-                        self.resolve_crate_relative_path(trait_path.span, segments, TypeNS)
-                    } else {
-                        self.resolve_module_relative_path(trait_path.span, segments, TypeNS)
-                    }.map(|binding| binding.span).unwrap_or(syntax_pos::DUMMY_SP)
-                };
-
-                if definition_site != syntax_pos::DUMMY_SP {
-                    err.span_label(definition_site,
-                                   &format!("type aliases cannot be used for traits"));
-                }
+                err.note(&format!("type aliases cannot be used for traits"));
             }
             err.emit();
             Err(true)

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -1810,6 +1810,8 @@ impl<'a> Resolver<'a> {
             if let Def::Trait(_) = path_res.base_def {
                 debug!("(resolving trait) found trait def: {:?}", path_res);
                 Ok(path_res)
+            } else if path_res.base_def == Def::Err {
+                Err(true)
             } else {
                 let mut err =
                     resolve_struct_error(self,

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -1807,41 +1807,40 @@ impl<'a> Resolver<'a> {
                                path_depth: usize)
                                -> Result<PathResolution, ()> {
         self.resolve_path(id, trait_path, path_depth, TypeNS).and_then(|path_res| {
-            if let Def::Trait(_) = path_res.base_def {
-                debug!("(resolving trait) found trait def: {:?}", path_res);
-                Ok(path_res)
-            } else if path_res.base_def == Def::Err {
-                Err(true)
-            } else {
-                let mut err =
-                    resolve_struct_error(self,
-                                  trait_path.span,
-                                  ResolutionError::IsNotATrait(&path_names_to_string(trait_path,
-                                                                                      path_depth)));
-
-                // If it's a typedef, give a note
-                if let Def::TyAlias(..) = path_res.base_def {
-                    let trait_name = trait_path.segments.last().unwrap().identifier.name;
-                    err.span_label(trait_path.span,
-                                   &format!("`{}` is not a trait", trait_name));
-
-                    let definition_site = {
-                        let segments = &trait_path.segments;
-                        if trait_path.global {
-                            self.resolve_crate_relative_path(trait_path.span, segments, TypeNS)
-                        } else {
-                            self.resolve_module_relative_path(trait_path.span, segments, TypeNS)
-                        }.map(|binding| binding.span).unwrap_or(syntax_pos::DUMMY_SP)
-                    };
-
-                    if definition_site != syntax_pos::DUMMY_SP {
-                        err.span_label(definition_site,
-                                       &format!("type aliases cannot be used for traits"));
-                    }
+            match path_res.base_def {
+                Def::Trait(_) => {
+                    debug!("(resolving trait) found trait def: {:?}", path_res);
+                    return Ok(path_res);
                 }
-                err.emit();
-                Err(true)
+                Def::Err => return Err(true),
+                _ => {}
             }
+
+            let mut err = resolve_struct_error(self, trait_path.span, {
+                ResolutionError::IsNotATrait(&path_names_to_string(trait_path, path_depth))
+            });
+
+            // If it's a typedef, give a note
+            if let Def::TyAlias(..) = path_res.base_def {
+                let trait_name = trait_path.segments.last().unwrap().identifier.name;
+                err.span_label(trait_path.span, &format!("`{}` is not a trait", trait_name));
+
+                let definition_site = {
+                    let segments = &trait_path.segments;
+                    if trait_path.global {
+                        self.resolve_crate_relative_path(trait_path.span, segments, TypeNS)
+                    } else {
+                        self.resolve_module_relative_path(trait_path.span, segments, TypeNS)
+                    }.map(|binding| binding.span).unwrap_or(syntax_pos::DUMMY_SP)
+                };
+
+                if definition_site != syntax_pos::DUMMY_SP {
+                    err.span_label(definition_site,
+                                   &format!("type aliases cannot be used for traits"));
+                }
+            }
+            err.emit();
+            Err(true)
         }).map_err(|error_reported| {
             if error_reported { return }
 

--- a/src/test/compile-fail/issue-3907.rs
+++ b/src/test/compile-fail/issue-3907.rs
@@ -11,14 +11,14 @@
 // aux-build:issue_3907.rs
 extern crate issue_3907;
 
-type Foo = issue_3907::Foo; //~ NOTE: type aliases cannot be used for traits
+type Foo = issue_3907::Foo;
 
 struct S {
     name: isize
 }
 
 impl Foo for S { //~ ERROR: `Foo` is not a trait
-    //~| `Foo` is not a trait
+                 //~| NOTE: type aliases cannot be used for traits
     fn bar() { }
 }
 

--- a/src/test/compile-fail/issue-5035.rs
+++ b/src/test/compile-fail/issue-5035.rs
@@ -13,4 +13,8 @@ type K = I;
 //~^ NOTE: aliases cannot be used for traits
 impl K for isize {} //~ ERROR: `K` is not a trait
 //~| is not a trait
+
+use ImportError; //~ ERROR unresolved
+impl ImportError for () {} // check that this is not an additional error (c.f. #35142)
+
 fn main() {}

--- a/src/test/compile-fail/issue-5035.rs
+++ b/src/test/compile-fail/issue-5035.rs
@@ -10,9 +10,8 @@
 
 trait I {}
 type K = I;
-//~^ NOTE: aliases cannot be used for traits
 impl K for isize {} //~ ERROR: `K` is not a trait
-//~| is not a trait
+                    //~| NOTE: aliases cannot be used for traits
 
 use ImportError; //~ ERROR unresolved
 impl ImportError for () {} // check that this is not an additional error (c.f. #35142)

--- a/src/test/ui/codemap_tests/two_files.stderr
+++ b/src/test/ui/codemap_tests/two_files.stderr
@@ -2,12 +2,9 @@ error[E0404]: `Bar` is not a trait
   --> $DIR/two_files.rs:16:6
    |
 16 | impl Bar for Baz { }
-   |      ^^^ `Bar` is not a trait
-   | 
-  ::: $DIR/two_files_data.rs
+   |      ^^^
    |
-15 | type Bar = Foo;
-   | --------------- type aliases cannot be used for traits
+   = note: type aliases cannot be used for traits
 
 error: cannot continue compilation due to previous error
 

